### PR TITLE
Move exponential_ from TH to Aten (CPU)

### PR DESCRIPTION
**Pull request by hl475**
Summary:
This diff will address https://github.com/pytorch/pytorch/issues/24699

We ask the input `lambda` to be >= 0 to be same as https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.exponential.html#numpy-random-exponential. This does not exist in the previous implementation.

Benchmark script
```
import torch
import torch.nn as nn
import time

torch.manual_seed(0)

def _time():
    return time.time()

device = "cpu"

#warm up
for n in [10, 100, 1000]:
    input = torch.randn(128, n, requires_grad=False, device=device)
    for i in range(1000):
        input.exponential_()

for n in [1, 10, 100, 1000]:
    fwd_t = 0
    input = torch.randn(128, n, requires_grad=False, device=device)
    for i in range(10000):
        t1 = _time()
        input.exponential_()
        t2 = _time()
        fwd_t = fwd_t + (t2 -t1)
    fwd_avg = fwd_t / 10000 * 1000
    print("input size(128, %d) forward time is %.4f (ms)." % (n, fwd_avg))
```
Output
```
================================================================================
Before the change, Program Output:
================================================================================
Higher priority item already registered, skipping registration of LINUX
input size(128, 1) forward time is 0.0129 (ms).
input size(128, 10) forward time is 0.1069 (ms).
input size(128, 100) forward time is 1.0409 (ms).
input size(128, 1000) forward time is 10.3827 (ms).
================================================================================
After the change, Program Output:
================================================================================
2020-01-22 10:52:06,406 fbpkg.fetch INFO: completed download of bento_kernel_pytorch:641
2020-01-22 10:52:06,407 fbpkg.fetch INFO: extracted bento_kernel_pytorch:641 to bento_kernel_pytorch
All packages downloaded successfully
Higher priority item already registered, skipping registration of LINUX
input size(128, 1) forward time is 0.0129 (ms).
input size(128, 10) forward time is 0.1065 (ms).
input size(128, 100) forward time is 1.0374 (ms).
input size(128, 1000) forward time is 10.3569 (ms).
================================================================================
```

Test Plan: Sandcastle and Github tests

Differential Revision: D19518700

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit fbbd55cc:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=fbbd55cc352407b3d04422ba3a59502ca2e9c5a8).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4319144) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89550513) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89550513))
<details>
<summary>
<code>Jan 22 14:11:41 RuntimeError: test_autograd failed!</code>
</summary>

```
Jan 22 14:11:40 Ran 1046 tests in 283.128s 
Jan 22 14:11:40  
Jan 22 14:11:40 FAILED (errors=1, skipped=13, expected failures=1) 
Jan 22 14:11:40  
Jan 22 14:11:40 Generating XML reports... 
Jan 22 14:11:41 Traceback (most recent call last): 
Jan 22 14:11:41   File "test/run_test.py", line 457, in <module> 
Jan 22 14:11:41     main() 
Jan 22 14:11:41   File "test/run_test.py", line 450, in main 
Jan 22 14:11:41     raise RuntimeError(message) 
Jan 22 14:11:41 RuntimeError: test_autograd failed! 
Jan 22 14:11:41 + cleanup 
Jan 22 14:11:41 + retcode=1 
Jan 22 14:11:41 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by hl475** [comment]
pytorch_macos_10_13_py3_test has been passed after retest.

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 1]
```diff
@@ -322,6 +322,17 @@ void bernoulli_mkl_kernel(Tensor &self, const double p, Generator* gen) {
 }
 #endif
 
+static void exponential_kernel(TensorIterator& iter, double lambda, Generator* gen) {
+  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "exponential_cpu", [&]() {
+    CPUGenerator* generator = get_generator_or_default<CPUGenerator>(gen, detail::getDefaultCPUGenerator());
+    std::lock_guard<std::mutex> lock(generator->mutex_);
+    cpu_serial_kernel(iter, [lambda, generator]() -> scalar_t {
+      at::exponential_distribution<double> exponential(lambda);
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp]: `at::exponential_distribution<double> exponential(lambda);` should be outside of loop statement

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]

**Comment by hl475** [comment]
address comments from @VitalyFedyunin

**Comment by hl475** [comment]
add more test coverage and explicitly comment for the case when lambda is 0.

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by zou3519** [comment]
The case where `lambda=0` should be different from `scale=0` in the numpy case, right?
numpy uses the `scale` parameter, which is equivalent to `1/lambda` in our API https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.exponential.html#numpy-random-exponential  https://pytorch.org/docs/stable/tensors.html#torch.Tensor.exponential_

This PR changed the behavior of `lambda=0` to return a tensor full of zeros, but it should really match the behavior of `numpy.random.exponential(scale=99999999, shape=10)` (some big number) and return a bit number.

Does that sound right,  @hl475 @VitalyFedyunin ?

**Comment by gchanan** [comment]
I think both scale and the result should be `inf`, i.e.:

```
>>> np.random.exponential(np.array(1.)/np.array(0.))
//anaconda2/envs/pytorch1.4/bin/ipython:1: RuntimeWarning: divide by zero encountered in true_divide
  #!//anaconda2/envs/pytorch1.4/bin/python
inf
```

**Pull request closed without merging**